### ISSUE TITLE
Stellar mass BHs, BNS, NSBH population analysis details

### ISSUE THREAD

**Issue by KrishnenduGit** [KrishnenduGit/CE_STM_CBC#3]

**Comment by sanikakhadkikar** [comment]
## Review Statement

**Reviewer**: @sanikakhadkikar  
**Code written by**: @shiksha1pandey  

**Date of review**: 2026-02-01  
**Git commit**: `ebc52b0`  
**Conda environment**: `/ligo/home/ligo.org/shiksha.pandey/.conda/envs/gwfast311/bin/python`

---

### 1. `mass_sampler.py`

**Description**: Generates mass samples from popsummary files.

#### Function-wise review

- [x] **`load_hyperparameters`**  
  Used to obtain median values of hyperparameters from the popsummary file.  
  The function is correctly written, but it may not be necessary if, as planned, hyperparameter values are taken directly from the GWTC-4 paper.

- [ ] **`generate_mass_ratio_samples`**  
  *Function did not pass review.* Comments below.

  ```python
  if source_type == 'bns':
      m_grid = np.linspace(1.0, 2.3, n_grid)
      q_grid = np.linspace(0.001, 1.0, n_grid)
  elif source_type == 'bbh':
      m_grid = np.geomspace(5.0, 500.0, n_grid)
      q_grid = np.linspace(0.001, 1.0, n_grid)
  elif source_type == 'nsbh':
      m_grid = np.geomspace(5.0, 500.0, n_grid)
      q_grid = np.geomspace(0.001, 0.5, n_grid)
  ```

  - `2.3`$M_{\odot}$ is not the maximum mass for the SFHO EoS used here and needs to be updated.
  - For NSBH systems, it is unclear how the mass grid is defined between the maximum NS mass and the minimum BH mass. Should this transition region be explicitly handled?
  - Similar concerns apply to the code pasted below. What is assumed to exist between the NS maximum mass and the BH minimum mass? Objects above the NS maximum mass should still be treated as BHs.

  ```python
  # Apply physical cuts. May need to change ns_max to 2.06 as that is the SFHO cut-off.
  ns_max, bh_min = 2.25, 5.0
  m2 = M * Q
  ```

  - `ns_max` should be updated to `2.06` $M_{\odot}$ for consistency with the SFHO EoS cutoff.
  - The choice of `q_min = 0.001` for the grid may need further justification. Flagging this for discussion with @KrishnenduGit.

---

### 2. `redshift_sampler.py`

**Description**: Generates redshift samples using the Madau–Dickinson star formation rate.

- [x] **`generate_md_redshift_samples`**  
  Uses GWforge to calculate the redshift distribution. Looks good to me.

---

### 3. `generate_tidal_params.py`

**Description**: Loads the SFHO EoS and creates the tidal deformability–mass relation.

- [ ] **`load_sfho_eos`**  
  Overall looks good.

  - The PyCompose portion of the code can be simplified for computational speed-up by saving the pressure and energy arrays to file and reusing them. Specifically:
    ```python
    pressure = eos_be.thermo['Q1'].flatten() * nb * mevfm3_gcm3 * cgs_to_geom
    energy = (eos_be.thermo['Q7'].flatten() + 1) * nb * m_n * mevfm3_gcm3 * cgs_to_geom
    ```
    This step only needs to be done once, as it can be saved as a text file.

  - For the interpolator:
    ```python
    self.interp = interp1d(
        m_grid[valid],
        lam_grid[valid],
        kind='cubic',
        bounds_error=False,
        fill_value=0.0
    )
    ```
    The `fill_value` for samples beyond the valid mass range should be set to `NaN`, since the tidal deformability cannot be computed where the EoS does not exist.

  - Apart from these points, the rest of the implementation looks good.

---

### 4. `create_injections.py`

**Description**: Creates injection samples for GW analysis.

- [x] Very cleanly written code with appropriate failure checks at each step. Looks great to me!

**Comment by KrishnenduGit** [comment]
@sanikakhadkikar and @sanikakhadkikar 
Thanks a lot for the notebooks and reviews. 

My comments are below: with the [commit](https://github.com/KrishnenduGit/CE_STM_CBC/commit/8a5a9d8ed0de8ff5823806a86fae9540315b7c92) 

- [`mass_sampler.py`](https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/mass_sampler.py)

1. As we discussed already, please do not generate hyper-parameters but use the inferred values from GWTC-4 rate and population paper. In other words, we do not need to use the function load_hyperparameters(pdb_file, model_class). In general, if a function is unused please remove it from the code to avoid any future confusion. 
2. The above change would automatically change https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/mass_sampler.py#L75 
3. As @sanikakhadkikar pointed out, ` if source_type == 'bns':
        m_grid = np.linspace(1.0, 2.3, n_grid)
        q_grid = np.linspace(0.001, 1.0, n_grid)` 

needs to be updated along with an update to 

`elif source_type == 'bbh':
        m_grid = np.geomspace(5.0, 500.0, n_grid)
        q_grid = np.linspace(0.001, 1.0, n_grid)
    elif source_type == 'nsbh':
        m_grid = np.geomspace(5.0, 500.0, n_grid)
        q_grid = np.geomspace(0.001, 0.5, n_grid)`

the NS_max should be changed following SHFo EoS, and the BH_min mass should be anything above NS_max mass value. q=0.001 is too high. We should restrict to q<0.1 imo. 

4. NS mass update to this line as well: https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/mass_sampler.py#L108 


- [`generate_tidal_params.py`](https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/generate_tidal_params.py)

1. We need a better documentation for this file. Especially, for definitions like this inside `load_sfho_eos`: 

`pressure = eos_be.thermo['Q1'].flatten() * nb * mevfm3_gcm3 * cgs_to_geom
    energy = (eos_be.thermo['Q7'].flatten() + 1) * nb * m_n * mevfm3_gcm3 * cgs_to_geom`

2. Also, it would good to have tests run on the mass range we agreed https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/generate_tidal_params.py#L162 

- [`redshift_sampler.py`](https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/redshift_sampler.py)


1. I suggest we take the redshift model directly from gwpopulation if possible: such as this line, https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/redshift_sampler.py#L2
2. All the hyper parameters used https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/redshift_sampler.py#L49, need to be verified with the GWTC-4 inference. In addition we should check how the new mass model impacts the redshift evolution and rate estimate. 

- [`create_injections.py`](https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/create_injections.py)

1. For the conversion functions, a proper documentation is required: see this [line](create_injections.py)
2. What are the properties of these [spin](https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/create_injections.py#L22) and [Tidal param conversion functions](https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/injection_sampler/create_injections.py#L31) do? 
3. I would really like to see a real instance of this file, with all inputs. So, it would be really great if you could add some details about the usage of this function. Especially on how the hyper parameters are passed.

**Comment by KrishnenduGit** [comment]
@sanikakhadkikar and @shiksha1pandey 
If we dont need the function load_hyperparameters(pdb_file, model_class): but we use the inferred hyper parameters from GWTC-4 (as we agreed, this is the best thing to do), can we all use this GIT repo as the central repo?

**Comment by KrishnenduGit** [comment]
- Reference for SHFo: See Tab. II of https://arxiv.org/pdf/1207.2184

**Comment by KrishnenduGit** [comment]
Following our decision to follow B7 and B8 of GWTC-4 astropop paper, I have reviewed the function used in https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/pop_models/full_stellar_mass_pop/mass_sampler_to_be_reviewed.py (actually the one used in https://github.com/KrishnenduGit/CE_STM_CBC/blob/main/pop_models/full_stellar_mass_pop/ce_stm_new.ipynb)
One comment I have is, if we correctly account for the normalization of this probability function. Perhaps this is automatically done within GWFOorge/GWPopulation?
Also, it would be good to plot m1, and m2 for BBH, BNS and NSBH cases
